File: wpilibc/src/test/native/cpp/smartdashboard/Mechanism2dTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/smartdashboard/Mechanism2d.hpp"

#include <memory>
#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "wpi/smartdashboard/MechanismLigament2d.hpp"
#include "wpi/telemetry/MockTelemetryBackend.hpp"
#include "wpi/telemetry/Telemetry.hpp"
#include "wpi/telemetry/TelemetryRegistry.hpp"
#include "wpi/units/angle.hpp"
#include "wpi/util/Color8Bit.hpp"

struct Mechanism2dTest {
  Mechanism2dTest() {
    wpi::telemetry::TelemetryRegistry::Reset();
    wpi::telemetry::TelemetryRegistry::RegisterBackend("", mock);
  }

  ~Mechanism2dTest() { wpi::telemetry::TelemetryRegistry::Reset(); }

  std::shared_ptr<wpi::telemetry::MockTelemetryBackend> mock =
      std::make_shared<wpi::telemetry::MockTelemetryBackend>();
};

TEST_CASE_METHOD(Mechanism2dTest, "Mechanism2dTest Canvas",
                 "[wpilibc][smartdashboard]") {
  wpi::Mechanism2d mechanism{5, 10};

  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto actions = mock->GetActions();
    REQUIRE(actions.size() == 3u);

    auto dims = mock->GetLastValue<std::vector<double>>("/mechanism/dims");
    REQUIRE(dims);
    REQUIRE(2u == dims->size());
    CHECK(5.0 == (*dims)[0]);
    CHECK(10.0 == (*dims)[1]);

    auto color = mock->GetLastValue<
        wpi::telemetry::MockTelemetryBackend::LogStringValue>(
        "/mechanism/backgroundColor");
    REQUIRE(color);
    CHECK("#000020" == color->value);
    mock->Clear();
  }

  mechanism.SetBackgroundColor({255, 255, 255});
  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto color = mock->GetLastValue<
        wpi::telemetry::MockTelemetryBackend::LogStringValue>(
        "/mechanism/backgroundColor");
    REQUIRE(color);
    CHECK("#FFFFFF" == color->value);
  }
}

TEST_CASE_METHOD(Mechanism2dTest, "Mechanism2dTest Root",
                 "[wpilibc][smartdashboard]") {
  wpi::Mechanism2d mechanism{5, 10};
  wpi::MechanismRoot2d* root = mechanism.GetRoot("root", 1, 2);
  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto pos =
        mock->GetLastValue<std::vector<double>>("/mechanism/root/position");
    REQUIRE(pos);
    REQUIRE(2u == pos->size());
    CHECK(1.0 == (*pos)[0]);
    CHECK(2.0 == (*pos)[1]);
    mock->Clear();
  }
  root->SetPosition(2, 4);
  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto pos =
        mock->GetLastValue<std::vector<double>>("/mechanism/root/position");
    REQUIRE(pos);
    REQUIRE(2u == pos->size());
    CHECK(2.0 == (*pos)[0]);
    CHECK(4.0 == (*pos)[1]);
  }
}

TEST_CASE_METHOD(Mechanism2dTest, "Mechanism2dTest Ligament",
                 "[wpilibc][smartdashboard]") {
  wpi::Mechanism2d mechanism{5, 10};
  wpi::MechanismRoot2d* root = mechanism.GetRoot("root", 1, 2);
  wpi::MechanismLigament2d* ligament = root->Append<wpi::MechanismLigament2d>(
      "ligament", 3, wpi::units::degrees<>{90}, 1,
      wpi::util::Color8Bit{255, 255, 255});
  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto angle = mock->GetLastValue<double>("/mechanism/root/ligament/angle");
    REQUIRE(angle);
    CHECK(ligament->GetAngle() == *angle);
    auto color = mock->GetLastValue<
        wpi::telemetry::MockTelemetryBackend::LogStringValue>(
        "/mechanism/root/ligament/color");
    REQUIRE(color);
    CHECK(ligament->GetColor().HexString() == color->value);
    auto length = mock->GetLastValue<double>("/mechanism/root/ligament/length");
    REQUIRE(length);
    CHECK(ligament->GetLength() == *length);
    auto weight = mock->GetLastValue<double>("/mechanism/root/ligament/weight");
    REQUIRE(weight);
    CHECK(ligament->GetLineWeight() == *weight);
    mock->Clear();
  }

  ligament->SetAngle(wpi::units::degrees<>{45});
  ligament->SetColor({0, 0, 0});
  ligament->SetLength(2);
  ligament->SetLineWeight(4);
  wpi::telemetry::Log("mechanism", mechanism);
  {
    auto angle = mock->GetLastValue<double>("/mechanism/root/ligament/angle");
    REQUIRE(angle);
    CHECK(ligament->GetAngle() == *angle);
    auto color = mock->GetLastValue<
        wpi::telemetry::MockTelemetryBackend::LogStringValue>(
        "/mechanism/root/ligament/color");
    REQUIRE(color);
    CHECK(ligament->GetColor().HexString() == color->value);
    auto length = mock->GetLastValue<double>("/mechanism/root/ligament/length");
    REQUIRE(length);
    CHECK(ligament->GetLength() == *length);
    auto weight = mock->GetLastValue<double>("/mechanism/root/ligament/weight");
    REQUIRE(weight);
    CHECK(ligament->GetLineWeight() == *weight);
  }
}

File: wpilibj/src/test/java/org/wpilib/smartdashboard/Mechanism2dTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.smartdashboard;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.wpilib.telemetry.MockTelemetryBackend;
import org.wpilib.telemetry.MockTelemetryBackend.LogStringValue;
import org.wpilib.telemetry.MockTelemetryBackend.LogStructValue;
import org.wpilib.telemetry.Telemetry;
import org.wpilib.telemetry.TelemetryRegistry;
import org.wpilib.util.Color8Bit;

class Mechanism2dTest {
  MockTelemetryBackend m_mock;

  @BeforeEach
  public void init() {
    m_mock = new MockTelemetryBackend();
    TelemetryRegistry.reset();
    TelemetryRegistry.registerBackend("", m_mock);
  }

  @AfterEach
  public void shutdown() {
    TelemetryRegistry.reset();
    m_mock.close();
  }

  @Test
  void testCanvas() {
    try (var mechanism = new Mechanism2d(5, 10)) {
      Telemetry.log("mechanism", mechanism);
      assertArrayEquals(
          new double[] {5, 10}, m_mock.getLastValue("/mechanism/dims", double[].class));
      assertEquals(
          "#000020",
          m_mock.getLastValue("/mechanism/backgroundColor", LogStringValue.class).value());

      mechanism.setBackgroundColor(new Color8Bit(255, 255, 255));
      Telemetry.log("mechanism", mechanism);
      assertEquals(
          "#FFFFFF",
          m_mock.getLastValue("/mechanism/backgroundColor", LogStringValue.class).value());
    }
  }

  @Test
  void testRoot() {
    try (var mechanism = new Mechanism2d(5, 10)) {
      final var root = mechanism.getRoot("root", 1, 2);
      Telemetry.log("mechanism", mechanism);
      assertArrayEquals(
          new double[] {1, 2}, m_mock.getLastValue("/mechanism/root/position", double[].class));

      root.setPosition(2, 4);
      Telemetry.log("mechanism", mechanism);
      assertArrayEquals(
          new double[] {2, 4}, m_mock.getLastValue("/mechanism/root/position", double[].class));
    }
  }

  @Test
  void testLigament() {
    try (var mechanism = new Mechanism2d(5, 10)) {
      var root = mechanism.getRoot("root", 1, 2);
      var ligament =
          root.append(new MechanismLigament2d("ligament", 3, 90, 1, new Color8Bit(255, 255, 255)));
      Telemetry.log("mechanism", mechanism);
      assertEquals(
          ligament.getAngle(),
          m_mock.getLastValue("/mechanism/root/ligament/angle", LogStructValue.class).value());
      assertEquals(
          ligament.getColor().toHexString(),
          m_mock.getLastValue("/mechanism/root/ligament/color", LogStringValue.class).value());
      assertEquals(
          ligament.getLength(),
          m_mock.getLastValue("/mechanism/root/ligament/length", Double.class));
      assertEquals(
          ligament.getLineWeight(),
          m_mock.getLastValue("/mechanism/root/ligament/weight", Double.class));

      ligament.setAngle(45);
      ligament.setColor(new Color8Bit(0, 0, 0));
      ligament.setLength(2);
      ligament.setLineWeight(4);
      Telemetry.log("mechanism", mechanism);
      assertEquals(
          ligament.getAngle(),
          m_mock.getLastValue("/mechanism/root/ligament/angle", LogStructValue.class).value());
      assertEquals(
          ligament.getColor().toHexString(),
          m_mock.getLastValue("/mechanism/root/ligament/color", LogStringValue.class).value());
      assertEquals(
          ligament.getLength(),
          m_mock.getLastValue("/mechanism/root/ligament/length", Double.class));
      assertEquals(
          ligament.getLineWeight(),
          m_mock.getLastValue("/mechanism/root/ligament/weight", Double.class));
    }
  }
}
